File: space-game/Data.SqlDataProvider/Classes/DataProvider.vb
Imports Core.Extensions
Imports Data.Classes
Imports Data.SqlDataProvider.Extensions
Imports Mapster

Namespace Classes
    Public Class DataProvider
        Implements IDataProvider

        Public Sub New()
        End Sub

        Public ReadOnly Property SolarSystem As List(Of CelestialObjectData) Implements IDataProvider.SolarSystem
            Get
                Dim result As New List(Of CelestialObjectData)
                Dim system = GetSolarSystem()
                If system.HasAny() Then
                    result.AddRange(From co In system Where Not co.PrimaryId.HasValue Select RecursiveCreateObject(co))
                End If
                Return result
            End Get
        End Property

        Private Function RecursiveCreateObject(co As CelestialObject) As CelestialObjectData
            Dim orbit As OrbitData = If(co.PrimaryId.HasValue, co.Adapt(Of OrbitData)(), Nothing)
            Dim physical As PhysicalData = co.Adapt(Of PhysicalData)()

            Dim primary = New CelestialObjectData(orbit, physical)

            If co.CelestialObject1.HasAny() Then
                For Each satellite In co.CelestialObject1
                    Dim s = RecursiveCreateObject(satellite)

                    If s IsNot Nothing Then
                        primary.AddSatellite(s)
                    End If
                Next
            End If

            Return primary
        End Function

        Private Function GetSolarSystem() As List(Of CelestialObject)
            Return GetSolarSystemByName("Solar System")
        End Function

        Private Function GetSolarSystemByName(name As String) As List(Of CelestialObject)
            If Not String.IsNullOrWhiteSpace(name) Then
                Using con As New SolarSystemEntities
                    con.Configuration.LazyLoadingEnabled = False
                    con.Configuration.ProxyCreationEnabled = False

                    Dim result = con.CelestialObjects.
                                Where(Function(o) o.SolarSystem.Name = name).
                                IncludeAllTables().
                                OrderBy(Function(p) p.PrimaryId).
                                ToList()

                    Return result
                End Using
            End If
            Return Nothing
        End Function

    End Class
End Namespace

File: space-game/Data.SqlDataProvider/Classes/DataProvider.cs
using Core.Extensions;
using Data.Classes;
using Data.SqlDataProvider.Extensions;
using Data.SqlDataProvider.Model;
using Mapster;
using System.Collections.Generic;
using System.Linq;

namespace Data.SqlDataProvider.Classes
{
    public class DataProvider : IDataProvider
    {
        public List<CelestialObjectData> SolarSystem
        {
            get
            {
                List<CelestialObjectData> result = new List<CelestialObjectData>();
                List<CelestialObject> system = GetSolarSystem();
                if (system.HasAny())
                {
                    result.AddRange(from co in system where !co.PrimaryId.HasValue select RecursiveCreateObject(co));
                }
                return result;
            }
        }

        private CelestialObjectData RecursiveCreateObject(CelestialObject co)
        {
            OrbitData orbit = co.PrimaryId.HasValue ? co.Adapt<OrbitData>() : null;
            PhysicalData physical = co.Adapt<PhysicalData>();

            CelestialObjectData primary = new CelestialObjectData(orbit, physical);

            if (co.CelestialObject1.HasAny())
            {
                foreach (CelestialObject satellite in co.CelestialObject1)
                {
                    CelestialObjectData s = RecursiveCreateObject(satellite);

                    if (s != null)
                    {
                        primary.AddSatellite(s);
                    }
                }
            }
            return primary;
        }

        private List<CelestialObject> GetSolarSystem()
        {
            return GetSolarSystemByName("Solar System");
        }

        private List<CelestialObject> GetSolarSystemByName(string name)
        {
            if (!string.IsNullOrWhiteSpace(name))
            {
                using (SolarSystemEntities con = new SolarSystemEntities())
                {
                    con.Configuration.LazyLoadingEnabled = false;
                    con.Configuration.ProxyCreationEnabled = false;

                    List<CelestialObject> result = con.CelestialObjects
                                    .Where(o => o.SolarSystem.Name == name)
                                    .IncludeAllTables()
                                    .OrderBy(p => p.PrimaryId).ToList();

                    return result;
                }
            }
            return null;
        }
    }
}
